File: tyhp-lang/phpstorm/src/unitTest/kotlin/com/tyhp/lang/debug/TyhpJsonTest.kt
package com.tyhp.lang.debug

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class TyhpJsonTest {
    @Test
    fun `reads xdebugProxy ports, sourcemap dir, ideKey, generateSourcemap, and output path`() {
        val snapshot = parseTyhpJsonProject(
            """
            {
                "build": { "generateSourcemap": true },
                "output": { "path": "out/" },
                "xdebugProxy": {
                    "idePort": 9010,
                    "xdebugPort": 9011,
                    "sourceMapDir": "./maps",
                    "ideKey": "tyhp"
                }
            }
            """.trimIndent(),
        )
        assertEquals(true, snapshot?.generateSourcemap)
        assertEquals("out/", snapshot?.outputPath)
        assertEquals(9010, snapshot?.xdebugProxy?.idePort)
        assertEquals(9011, snapshot?.xdebugProxy?.xdebugPort)
        assertEquals("./maps", snapshot?.xdebugProxy?.sourceMapDir)
        assertEquals("tyhp", snapshot?.xdebugProxy?.ideKey)
    }

    @Test
    fun `generateSourcemap defaults to false when omitted`() {
        val snapshot = parseTyhpJsonProject("""{ "output": { "path": "build/" } }""")
        assertEquals(false, snapshot?.generateSourcemap)
        assertEquals("build/", snapshot?.outputPath)
        assertNull(snapshot?.xdebugProxy)
    }

    @Test
    fun `accepts numeric ports encoded as strings`() {
        val snapshot = parseTyhpJsonProject(
            """{ "xdebugProxy": { "idePort": "9111", "xdebugPort": "9222" } }""",
        )
        assertEquals(9111, snapshot?.xdebugProxy?.idePort)
        assertEquals(9222, snapshot?.xdebugProxy?.xdebugPort)
    }

    @Test
    fun `ignores null sourceMapDir and empty strings`() {
        val snapshot = parseTyhpJsonProject(
            """{ "xdebugProxy": { "sourceMapDir": null, "ideKey": "" } }""",
        )
        assertNull(snapshot?.xdebugProxy?.sourceMapDir)
        assertNull(snapshot?.xdebugProxy?.ideKey)
    }

    @Test
    fun `returns null for invalid JSON`() {
        assertNull(parseTyhpJsonProject("{"))
        assertNull(parseTyhpJsonProject("[]"))
    }
}

File: tyhp-lang/vscode/src/debug/tyhpJson.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import { parseTyhpJsonProject } from "./tyhpJson";

test("reads xdebugProxy ports, sourcemap dir, ideKey, generateSourcemap, and output.path", () => {
    const snapshot = parseTyhpJsonProject(`{
        "build": { "generateSourcemap": true },
        "output": { "path": "out/" },
        "xdebugProxy": {
            "idePort": 9010,
            "xdebugPort": 9011,
            "sourceMapDir": "./maps",
            "ideKey": "tyhp"
        }
    }`);
    assert.deepEqual(snapshot, {
        generateSourcemap: true,
        outputPath: "out/",
        xdebugProxy: {
            idePort: 9010,
            xdebugPort: 9011,
            sourceMapDir: "./maps",
            ideKey: "tyhp",
        },
    });
});

test("generateSourcemap defaults to false when omitted", () => {
    const snapshot = parseTyhpJsonProject(`{ "output": { "path": "build/" } }`);
    assert.equal(snapshot?.generateSourcemap, false);
    assert.equal(snapshot?.outputPath, "build/");
    assert.equal(snapshot?.xdebugProxy, undefined);
});

test("accepts numeric ports encoded as strings", () => {
    const snapshot = parseTyhpJsonProject(`{
        "xdebugProxy": { "idePort": "9111", "xdebugPort": "9222" }
    }`);
    assert.equal(snapshot?.xdebugProxy?.idePort, 9111);
    assert.equal(snapshot?.xdebugProxy?.xdebugPort, 9222);
});

test("ignores null sourceMapDir and empty strings", () => {
    const snapshot = parseTyhpJsonProject(`{
        "xdebugProxy": { "sourceMapDir": null, "ideKey": "" }
    }`);
    assert.equal(snapshot?.xdebugProxy?.sourceMapDir, undefined);
    assert.equal(snapshot?.xdebugProxy?.ideKey, undefined);
});

test("returns undefined for invalid JSON", () => {
    assert.equal(parseTyhpJsonProject("{"), undefined);
    assert.equal(parseTyhpJsonProject("[]"), undefined);
});
